app/src/keymap.c

```diff
@@ -174,22 +174,29 @@ int invoke_locally(struct zmk_behavior_binding *binding, struct zmk_behavior_bin
 
 int zmk_keymap_apply_position_state(uint8_t source, int layer, uint32_t position, bool pressed,
                                     int64_t timestamp) {
-    // We want to make a copy of this, since it may be converted from
-    // relative to absolute before being invoked
-    struct zmk_behavior_binding binding = zmk_keymap[layer][position];
-    const struct device *behavior;
+    struct zmk_behavior_binding *binding = &zmk_keymap[layer][position];
     struct zmk_behavior_binding_event event = {
         .layer = layer,
         .position = position,
         .timestamp = timestamp,
+        .source = source,
     };
 
-    LOG_DBG("layer: %d position: %d, binding name: %s", layer, position, binding.behavior_dev);
+    LOG_DBG("layer: %d position: %d, binding name: %s", layer, position, binding->behavior_dev);
+
+    return zmk_invoke_behavior_binding(binding, event, pressed);
+}
+
+int zmk_invoke_behavior_binding(const struct zmk_behavior_binding *src_binding,
```

Comment:
If we're going to start having external callers of this, I'm not *convinced* this should stay in keymap, instead of going somewhere like `app/src/behavior.c`.

Reply:
Makes sense, I'll move this to behavior.c and the corresponding header.

Reply:
I moved the function and the `invoke_locally` method, I think it is better indeed.